doc/ocaml-ecosystem.rst

```diff
@@ -0,0 +1,89 @@
+on top of the compilers to perform tasks such as producing an executable from
+compiled object files and external libraries.
+
+Opam: a Collection of Software Projects
+---------------------------------------
+
+Opam is a package manager. It is used to determine which packages are
+necessary, and how to fetch and build them. Packages can contain libraries,
+executables, and other kinds of files.
+
+The notion of version is specific to opam. If your project uses a function
+named ``Png.read_file`` but this function has been added only in version
+``1.2.0`` of that package, opam needs to know about it.
+
+Opam manages collections of installed packages, called switches. Using your
+project's dependencies (names and version constraints), it is able to create a
+switch that you'll be using to develop your project.
+
+Public definition of packages are available in a database called
+``opam-repository`` which is maintained as a public Git repository. Publishing a
+package on opam (to make sure that external users can use your project)
+consists in adding its definition to ``opam-repository``.
+
+Dune: Giving Structure to Your Source Tree
+------------------------------------------
+
+Dune is a build system. It is used to orchestrate the compilation of source
+files into executables and libraries.
+
+Assuming you have a development switch set up, you communicate to Dune about how your
+project is organized in terms of executables, libraries, and tests. It is then able to assemble the source files of your projects, with the dependencies installed in an opam switch, to create compiled assets for your project.
+
+How Dune Integrates with the Ecosystem
+--------------------------------------
+
+Dune is designed to integrate with this set of tools:
+
+- By knowing how the compilers operate, it knows which build commands should be
+  re-executed if some source files change.
+- It outputs metadata like dependency information into ``META`` files that
+  Findlib is able to use. This ensures that even if a project does not use Dune, it
+  can use a library that has been produced by Dune. Conversely, it can read
+  these files to determine dependency information for dependencies that have
+  not been produced by Dune.
+- It is able to generate opam files with filenames consistents with how opam
+  looks for them. The generated files use build commands that make use of the
+  ``@install`` and ``@runtest`` alias so that the Dune abstractions map to the
+  opam ones.
+
+Dune is Opinionated
+-------------------
+
+As described above, the OCaml ecosystem does not have a centralized toolchain.
+Units such as modules, libraries, and packages operate at different levels, and
+the relation between these can be confusing to users.
+
+Dune tries to simplify the picture by reducing the difference between these
+objects:
+
+- By default, a library will only expose a single top-level module named after
+  the library.
```

Comment:
Perhaps mention: This is known as a wrapped library. And then link to the reference manual.